<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:
- **TASK 1:** Mark all launch sites on a map
- **TASK 2:** Mark the success/failed launches for each site on the map
- **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
#!pip3 install folium
#!pip3 install wget
#!pip3 install pandas

In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site. 


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

100% [................................................................................] 7710 / 7710

In [5]:
spacex_df.shape

(56, 13)

In [6]:
spacex_df

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
0               1  2010-06-04   18:45:00  F9 v1.0  B0003   CCAFS LC-40   
1               2  2010-12-08   15:43:00  F9 v1.0  B0004   CCAFS LC-40   
2               3  2012-05-22    7:44:00  F9 v1.0  B0005   CCAFS LC-40   
3               4  2012-10-08    0:35:00  F9 v1.0  B0006   CCAFS LC-40   
4               5  2013-03-01   15:10:00  F9 v1.0  B0007   CCAFS LC-40   
5               7  2013-12-03   22:41:00         F9 v1.1   CCAFS LC-40   
6               8  2014-01-06   22:06:00         F9 v1.1   CCAFS LC-40   
7               9  2014-04-18   19:25:00         F9 v1.1   CCAFS LC-40   
8              10  2014-07-14   15:15:00         F9 v1.1   CCAFS LC-40   
9              11  2014-08-05    8:00:00         F9 v1.1   CCAFS LC-40   
10             12  2014-09-07    5:00:00   F9 v1.1 B1011   CCAFS LC-40   
11             13  2014-09-21    5:52:00   F9 v1.1 B1010   CCAFS LC-40   
12             14  2015-01-10    9:47:00   F9 v1.1 B1012   CCAFS LC-40   
13             15  2015-02-11   23:03:00   F9 v1.1 B1013   CCAFS LC-40   
14             16  2015-03-02    3:50:00   F9 v1.1 B1014   CCAFS LC-40   
15             17  2015-04-14   20:10:00   F9 v1.1 B1015   CCAFS LC-40   
16             18  2015-04-27   23:03:00   F9 v1.1 B1016   CCAFS LC-40   
17             19  2015-06-28   14:21:00   F9 v1.1 B1018   CCAFS LC-40   
18             20  2015-12-22    1:29:00     F9 FT B1019   CCAFS LC-40   
19             22  2016-03-04   23:35:00     F9 FT B1020   CCAFS LC-40   
20             23  2016-04-08   20:43:00   F9 FT B1021.1   CCAFS LC-40   
21             24  2016-05-06    5:21:00     F9 FT B1022   CCAFS LC-40   
22             25  2016-05-27   21:39:00   F9 FT B1023.1   CCAFS LC-40   
23             26  2016-06-15   14:29:00     F9 FT B1024   CCAFS LC-40   
24             27  2016-07-18    4:45:00   F9 FT B1025.1   CCAFS LC-40   
25             28  2016-08-14    5:26:00     F9 FT B1026   CCAFS LC-40   
26              6  2013-09-29   16:00:00  F9 v1.1  B1003   VAFB SLC-4E   
27             21  2016-01-17   18:42:00   F9 v1.1 B1017   VAFB SLC-4E   
28             29  2017-01-14   17:54:00   F9 FT B1029.1   VAFB SLC-4E   
29             37  2017-06-25   20:25:00   F9 FT B1036.1   VAFB SLC-4E   
30             40  2017-08-24   18:51:00   F9 FT B1038.1   VAFB SLC-4E   
31             42  2017-10-09   12:37:00   F9 B4 B1041.1   VAFB SLC-4E   
32             46  2017-12-23    1:27:00  F9 FT  B1036.2   VAFB SLC-4E   
33             49  2018-02-22   14:17:00  F9 FT  B1038.2   VAFB SLC-4E   
34             51  2018-03-30   14:14:00  F9 B4  B1041.2   VAFB SLC-4E   
35             55  2018-05-22   19:47:58  F9 B4  B1043.2   VAFB SLC-4E   
36             30  2017-02-19   14:39:00   F9 FT B1031.1    KSC LC-39A   
37             31  2017-03-16    6:00:00     F9 FT B1030    KSC LC-39A   
38             32  2017-03-30   22:27:00  F9 FT  B1021.2    KSC LC-39A   
39             33  2017-05-01   11:15:00   F9 FT B1032.1    KSC LC-39A   
40             34  2017-05-15   23:21:00     F9 FT B1034    KSC LC-39A   
41             35  2017-06-03   21:07:00   F9 FT B1035.1    KSC LC-39A   
42             36  2017-06-23   19:10:00  F9 FT  B1029.2    KSC LC-39A   
43             38  2017-07-05   23:38:00     F9 FT B1037    KSC LC-39A   
44             39  2017-08-14   16:31:00   F9 B4 B1039.1    KSC LC-39A   
45             41  2017-09-07   14:00:00   F9 B4 B1040.1    KSC LC-39A   
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53      

Now, you can take a look at what are the coordinates for each site.


In [7]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [8]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example, 


In [9]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle. 


Now, let's add a circle for each launch site in data frame `launch_sites`


_TODO:_  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [10]:
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

In [11]:
# Initial the map
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=3)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label

marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    ).add_to(site_map)

#Add the equator line

folium.PolyLine(
    locations=[[0,-180],[0,180]],
    color='blue',
    weight=2,
    tooltip='Equator latitude 0°'
).add_to(site_map)

# Iterate over the df with the coordinates adding
#1. A circle marker for each site
#2. A vertical line to connect to the equator line
#3. Tooltips showing the distance to the equator
for _, row in launch_sites_df.iterrows():
    lat = row['Lat']
    lon = row['Long']
    name = row['Launch Site']

    folium.map.Marker(
    (lat,lon),
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html=f'<div style="font-size: 12; color:#d35400;"><b>{name}'
        )
    ).add_to(site_map)

#approximate distance from the equator
    #1 degree latitude, aprox 111 km
    distance_km=abs(lat)*111

     # Tooltip text for markers
    tooltip_text = (
        f"<b>{name}</b><br>"
        f"Distance to Equator: {distance_km:.1f} km"
    )

    # Create circle for the launch sites
    folium.Circle(
        location=[lat, lon],
        radius=1000,
        color='#d35400',
        fill=True,
        fill_color='#d35400',
        tooltip=tooltip_text
    ).add_to(site_map)

    folium.PolyLine(
        locations=[[lat, lon], [0, lon]],
        color='green',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_km:.1f} km from the Equator"
    ).add_to(site_map)
site_map


The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:
- Are all launch sites in proximity to the Equator line?
- Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


<h1>Answers</h1>
Are all launch sites in proximity to the Equator line?

Even though the launch sites are located more than 3000 km from the equator, they are still positioned at relative low latitudes on the coasts. This provides additional velocity boost during launches-.



# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [12]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records. 
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [13]:
marker_cluster = MarkerCluster()


_TODO:_ Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [14]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red


In [15]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

/tmp/ipykernel_31909/572784628.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)


Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

_TODO:_ For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [16]:
spacex_df[['Lat','Long','class']].isnull().sum()


Lat      0
Long     0
class    0
dtype: int64

In [17]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    lat = record['Lat']
    lon = record['Long']
    color = record['marker_color']
    site = record['Launch Site']
    result = "Success" if record['class'] == 1 else "Failure"
    # TODO: Create and add a Marker cluster to the site map
    marker = folium.Marker(
        location=[lat, lon],
        popup=f"{site} – {result}",
        icon=folium.Icon(
            color=color,      # borde del marcador
            icon_color='white' # color del icono
        )
    )
    # marker = folium.Marker(...)
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [18]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [19]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

_TODO:_ Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [20]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

In [21]:
west_coast=['VAFB SLC-4E']
east_coast=['CCAFS LC-40','CCAFS SLC-40','KSC LC-39A']

In [22]:


coastline_florida_lat = 28.56418
coastline_florida_lon = -80.56808
coastline_vandenberg_lat =34.63439
coastline_vandenberg_lon =-120.62562
distance_coastline_west = calculate_distance(34.632834, -120.610745, coastline_vandenberg_lat, coastline_vandenberg_lon)
distance_coastline_east = calculate_distance(28.562302,-80.577356, coastline_florida_lat, coastline_florida_lon)
print(f'the distance between the launch sites to the closes costline for weast and east are respectivly:{distance_coastline_west},{distance_coastline_east}')

the distance between the launch sites to the closes costline for weast and east are respectivly:1.3723212981600192,0.929955730125984


In [23]:
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

_TODO:_ After obtained its coordinate, create a `folium.Marker` to show the distance


In [24]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

In [25]:
folium.Marker(
    [coastline_vandenberg_lat, coastline_vandenberg_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_west),
    )
).add_to(site_map)
folium.Marker(
    [coastline_florida_lat, coastline_florida_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_east),
    )
).add_to(site_map)
folium.PolyLine(
        locations=[[34.632834,-120.610745], [coastline_vandenberg_lat, coastline_vandenberg_lon]],
        color='red',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_west}km from the closest coastline"
        ).add_to(site_map)
folium.PolyLine(
        locations=[[28.563197, -80.576820], [coastline_florida_lat, coastline_florida_lon]],
        color='green',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_east}km from the closest coastline"
        ).add_to(site_map)

site_map

_TODO:_ Draw a `PolyLine` between a launch site to the selected coastline point


In [26]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
#site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


_TODO:_ Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [27]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site


In [28]:


Florida_railway_lat = 28.56269
Florida_railway_lon = -80.58712
Florida_highway_lat = 28.56331
Florida_highway_lon = -80.57078
Florida_cape_canaveral_lat = 28.39321
Florida_cape_canaveral_lon = -80.60112
West_railway_lat = 34.63406
West_railway_lon = -120.62456
West_highway_lat = 34.69082
West_highway_lon = -120.45838
West_lompoc_lat = 34.63886
West_lompoc_lon = -120.45509

distance_coastline_west_railway = calculate_distance(34.632834, -120.610745, West_railway_lat, West_railway_lon)
distance_coastline_west_highway = calculate_distance(34.632834, -120.610745, West_highway_lat, West_highway_lon)
distance_coastline_west_city = calculate_distance(34.632834, -120.610745, West_lompoc_lat, West_lompoc_lon)


distance_coastline_east_railway = calculate_distance(28.562302,-80.577356, Florida_railway_lat, Florida_railway_lon)
distance_coastline_east_highway = calculate_distance(28.562302,-80.577356, Florida_highway_lat, Florida_highway_lon)
distance_coastline_east_city = calculate_distance(28.562302,-80.577356, Florida_cape_canaveral_lat, Florida_cape_canaveral_lon)


print(f'the distance between the launch site VAFB SLC-4E to the closest railway,highway, and city in the west are {distance_coastline_west_railway},{distance_coastline_west_highway},{distance_coastline_west_city} respectivly:')

print(f'the distance between the launch site CCAFS LC-40 to the closest railway,highway, and city in the west are {distance_coastline_east_railway},{distance_coastline_east_highway},{distance_coastline_east_city} respectivly:')

the distance between the launch site VAFB SLC-4E to the closest railway,highway, and city in the west are 1.2716875526886735,15.359550914135237,14.260959827258077 respectivly:
the distance between the launch site CCAFS LC-40 to the closest railway,highway, and city in the west are 0.9548476358995944,0.6521359948733605,18.95104309156464 respectivly:


In [29]:
folium.Marker(
    [West_railway_lat, West_railway_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_west_railway),
    )
).add_to(site_map)
folium.Marker(
    [West_highway_lat, West_highway_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_west_highway),
    )
).add_to(site_map)
folium.Marker(
    [West_lompoc_lat, West_lompoc_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_west_city),
    )
).add_to(site_map)
folium.Marker(
    [Florida_railway_lat, Florida_railway_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_east_railway),
    )
).add_to(site_map)
folium.Marker(
    [Florida_highway_lat, Florida_highway_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_east_highway),
    )
).add_to(site_map)
folium.Marker(
    [Florida_cape_canaveral_lat, Florida_cape_canaveral_lon],
    icon=DivIcon(
    icon_size=(20,20),
    icon_anchor=(0,0),
    html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline_east_city),
    )
).add_to(site_map)

#lines for distances in the west of railway, highway, and city to launch site VAFB SLC-4E

folium.PolyLine(
        locations=[[34.632834,-120.610745], [West_railway_lat, West_railway_lon]],
        color='red',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_west_railway}km from the closest railway from the launch site VAFB SLC-4E"
        ).add_to(site_map)
folium.PolyLine(
        locations=[[34.632834,-120.610745], [West_highway_lat, West_highway_lon]],
        color='red',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_west_highway}km from the closest highway from the launch site VAFB SLC-4E"
        ).add_to(site_map)
folium.PolyLine(
        locations=[[34.632834,-120.610745], [West_lompoc_lat, West_lompoc_lon]],
        color='red',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_west_city}km from the closest city from the launch site VAFB SLC-4E"
        ).add_to(site_map)


#lines for distances in the west of railway, highway, and city to launch site CCAFS LC-40


folium.PolyLine(
        locations=[[28.563197, -80.576820], [Florida_railway_lat, Florida_railway_lon]],
        color='green',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_east_railway}km from the closest railway from the launch site CCAFS LC-40"
        ).add_to(site_map)
folium.PolyLine(
        locations=[[28.563197, -80.576820], [Florida_highway_lat, Florida_highway_lon]],
        color='green',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_east_highway}km from the closest highway from the launch site CCAFS LC-40"
        ).add_to(site_map)
folium.PolyLine(
        locations=[[28.563197, -80.576820], [Florida_cape_canaveral_lat, Florida_cape_canaveral_lon]],
        color='green',
        weight=1,
        opacity=0.7,
        tooltip=f"{distance_coastline_east_city}km from the closest city from the launch site CCAFS LC-40"
        ).add_to(site_map)

site_map

In [30]:
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

After you plot distance lines to the proximities, you can answer the following questions easily:
- Are launch sites in close proximity to railways?
- Are launch sites in close proximity to highways?
- Are launch sites in close proximity to coastline?
- Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


<h1>Answers</h1>


<b>Are launch sites in close proximity to railways?</b>
Yes. Both launch sites are located very close to railways.

VAFB SLC-4E is approximately 1.2 km from the nearest railway.

CCAFS LC-40 is less than 1 km (0.9 km) from the nearest railway.
This proximity facilitates the transportation of heavy equipment and rocket components.

<b>Are launch sites in close proximity to highways?</b>
Yes. Launch sites are also located close to highways.

VAFB SLC-4E is approximately 15.3 km from the closest highway.

CCAFS LC-40 is approximately 0.6 km from the closest highway.
This allows efficient ground access and logistical support.

<b>Are launch sites in close proximity to the coastline?</b>
Yes. Both launch sites are located very close to the coastline.

The distance to the closest coastline is approximately 1.3 km for the west coast site (VAFB).

The distance is approximately 0.9 km for the east coast site (CCAFS).
This proximity is likely due to safety reasons, allowing rocket stages and debris to fall into the ocean.

<b>Do launch sites keep a certain distance away from cities?</b>
Yes. Launch sites are located far from major cities.

VAFB SLC-4E is approximately 14.2 km from the closest city, Lompoc.

CCAFS LC-40 is approximately 18.9 km from the closest city, Cape Canaveral.
This helps minimize risks to populated areas in case of launch failures.









# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/)


### Other Contributors


Joseph Santarcangelo


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2021-05-26|1.0|Yan|Created the initial version|


Copyright © 2021 IBM Corporation. All rights reserved.
